File: server/models/order_item.js
const { Schema, model } = require('mongoose');

const orderItemSchema = Schema({
  product: {
    type: Schema.Types.ObjectId,
    ref: 'Product',
    required: true,
  },
  // Historical Data Denormalization: When an order is created, we could denormalize some information from
  // the product (like name, image, price) into the order itself. This way, even if the product is deleted later,
  // the relevant information is preserved in the order document.
  productName: { type: String, required: true },
  productImage: { type: String, required: true },
  productPrice: { type: Number, required: true },
  quantity: { type: Number, default: 1 },
  selectedSize: String,
  selectedColour: String,
});

exports.OrderItem = model('OrderItem', orderItemSchema);

File: Mobile/lib/src/order/domain/entities/order_item.dart
import 'dart:ui';

import 'package:equatable/equatable.dart';

class OrderItem extends Equatable {
  const OrderItem({
    required this.id,
    required this.productId,
    required this.productName,
    required this.productImage,
    required this.productPrice,
    required this.quantity,
    this.selectedSize,
    this.selectedColour,
  });

  const OrderItem.empty()
      : id = "Test String",
        productId = "Test String",
        productName = "Test String",
        productImage = "Test String",
        productPrice = 1.0,
        quantity = 1,
        selectedSize = "Test String",
        selectedColour = null;

  final String id;
  final String productId;
  final String productName;
  final String productImage;
  final double productPrice;
  final int quantity;
  final String? selectedSize;
  final Color? selectedColour;

  @override
  List<dynamic> get props => [
        id,
        productId,
        productName,
        productImage,
        productPrice,
        quantity,
        selectedSize,
        selectedColour,
      ];
}
